### ISSUE TITLE
Using tensorflow and pytorch layers

### ISSUE THREAD

**Issue by datnamer** [FluxML/Flux.jl#521]
There are some basic and a long tail of pytorch and tensorflow packages which are unavailable in flux and this  might preclude my use of flux (unfortunately) for a couple of upcoming projects. 

How feasible would it be to somehow hook into either the python or C/C++apis of tensorflow or pytorch to use their layers in models with some of the more basic data flow? Obviously at the cost of some computational efficiency and flexibility. 

Can ONNX help? Also I  know there was some initial work on translating to different backends using a DAG...

**Comment by jekbradbury** [comment]
There are a couple options here. If the overhead isn't too big, you should be able to use PyCall to invoke PyTorch or TF libraries in Python, and pass the tensor values as CUDA pointers (there's definitely a little bit of glue code needed here). We can also create Julia bindings for PyTorch, or improve/modernize the Julia bindings for TensorFlow, but it's unlikely that would help if the goal is to be able to use external libraries unmodified.

I'm curious what packages you're interested in using!

**Comment by datnamer** [comment]
@jekbradbury  Thanks so much for your reply.

I want to do stuff with various advanced embedding and transfer learning models like BERT, parts of Allen NLP, some network embedding algorithms and graph CNNs etc.

However, there are some parts that would be better implemented in Julia (aside from the comfort of coding). I'm still a novice and I don't know if  I should spend the time trying to reimplement and validate/ troubleshoot the pre existing code. 

For CV side, it would be things like learning optical flow.

Would the proposed solution allow for end to end learning with foreign layers as well, or just inference?  

If so, I'd be happy to submit a PR with some direction (though I can't code in C or cuda).

**Comment by jekbradbury** [comment]
Getting BERT running in Flux, including TPU support and the ability to perform inference or finetuning using the pre-trained model checkpoints Google released, is something I intend to work on over the next few weeks.

For the other things: graph NNs would be a great fit for reimplementing in Flux; AllenNLP (and pytext, torchtext, tf.data, etc.) have data-related APIs that don't yet exist in Julia (but that's orthogonal to training/autodiff/Flux).

**Comment by datnamer** [comment]
> Getting BERT running in Flux, including TPU support and the ability to perform inference or finetuning using the pre-trained model checkpoints Google released, is something I intend to work on over the next few weeks.

That would be a big help! Looking forward. 

> have data-related APIs that don't yet exist in Julia (but that's orthogonal to training/autodiff/Flux).
Maybe that's part of the bloat that makes it seems intimidating. 

Perhaps I'll have a go then.

**Comment by MikeInnes** [comment]
Depending on what you need, one option might be to wrap ATen (PyTorch's C++ backend). This is what NimTorch did, and it seems like it was pretty successful as a way to get up and running. You could generate both the wrappers and gradient definitions hooked into Flux, and then it would interoperate with existing Flux layers.

In any case, feel free to open issues for any specific things that are missing (even if it's things like data loaders that may be outside of Flux's direct scope). Someone might be able to point you the right way, or at worst it gives us things to prioritise for work in the new year.

**Comment by datnamer** [comment]
> Depending on what you need, one option might be to wrap ATen (PyTorch's C++ backend)

This would only work for low level Cuda defined kernels/layers right? Not high level layers defined from python? 

What about pytorch JIT-_ONNX-> Flux? 

Any I'll open up some issues, thanks.  Looking forward to contributing.

**Comment by XVilka** [comment]
> For the other things: graph NNs would be a great fit for reimplementing in Flux; AllenNLP (and pytext, torchtext, tf.data, etc.) have data-related APIs that don't yet exist in Julia (but that's orthogonal to training/autodiff/Flux).

@jekbradbury yes, indeed it is a much wanted feature: https://github.com/FluxML/Flux.jl/issues/521

**Comment by darsnack** [comment]
I would say that this can safely be closed due to PyCall.jl and Torch.jl. And a lot of the use-cases mentioned (like BERT) are now available in Transformers.jl. Please re-open this if there is more to be addressed.
